docs/enhancements/add-external-providers.md

```diff
@@ -0,0 +1,202 @@
+define a clear interface which each provider can implement and maintain.
+
+### Goals
+
+- Define a clear way to add a specific platform provider to the assisted installer.
+- Migrate all existing supported providers to implement the same unified interface.
+
+### Non-Goals
+
+- Support IPI installations.
+- Integrating any relevant [Openshift installer types](https://github.com/openshift/installer/tree/master/pkg/types)
+into the assisted installer code - this will require a large refactor of the assisted installer code.
+- Remove Baremetal from not appropriate struct/function names.
+- Adding Support for cloud installations (AWS/GCP/Azure).
+
+## Proposal
+
+- Create a Provider interface, and hooks in the assisted installer code which separates the assisted installer core code
+ from the provider-specific code.
+- Add clear documentation for each provider specific function.
+- Document steps to add a provider to the assisted installer.
+
+### User Stories
+
+#### Story 1
+
+As an Openshift developer working on provider X I want to add support for installing Openshift with assisted installer
+on top of platform X.
+
+#### Story 2
+
+As an Openshift developer working on the assisted installer team I want to let providers hook into the
+assisted installer flow natural without having to worry about provider specific implementation.
+
+### Implementation Details/Notes/Constraints
+
+#### Changes to the assisted-service project:
+
+##### Step 1: Detect platform specific logic
+
+Go over the assisted installer flow and detect parts of the code which require provider specific logic.
+This step is based mostly on the work that already been done on the vSphere provider addition (MGMT-7067).
+
+##### Step 2: Create a Go interface
+
+We should create a Go interface that will contain each provider specific logic in a function
+for each provider to implement.
+
+Prototype:
+
+```go
+// Provider contains functions which are required to support installing on a specific platform.
+type Provider interface {
+    // Name returns the name of the platform.
+    Name() models.PlatformType
+    // AddPlatformToInstallConfig adds the provider platform to the installconfig platform field,
+    // sets platform fields from values within the cluster model.
+    AddPlatformToInstallConfig(cfg *installcfg.InstallerConfigBaremetal, cluster *common.Cluster) error
+    // SetPlatformValuesInDBUpdates updates the `updates` data structure with platform specific values
+    SetPlatformValuesInDBUpdates(platformParams *models.Platform, updates map[string]interface{}) error
+    // SetPlatformUsages uses the usageApi to update platform specific usages
+    SetPlatformUsages(platformParams *models.Platform, usages map[string]models.Usage, usageApi usage.API) error
+    // AreHostsSupported checks if the provider supports the hosts from values discovered in the inventory data
+    AreHostsSupported(hostInventories []*models.Inventory) (bool, error)
```

Comment:
1. I think a validation for a single host could be easier to manipulate and run in a loop if necessary instead of just supporting arrays.
2. I think the wrapping `models.Host` for the Inventory is better since we can add more business logic metadata to the host without it being detected by the agent and as such sent as part of the Inventory model.